### DISCUSSION TITLE
Manage tags improvements

### DISCUSSION THREAD

**Discussion by tedo-develop** [shlinkio/discussions#73] [Ideas]
Hello Acelaya,

our employees have some new suggestions for improvements.
I hope that they are interesting for shlink.

**Long loading time**
The page loads very long if you have many tags with statistics. 
This "problem" occurs also if you click the back button from visit page.

![Screenshot 2021-12-13 at 10 04 44](https://user-images.githubusercontent.com/39614855/145792054-e11751e5-e521-4bdb-a3a1-fda8f48d80be.png)

There is also a problem with site crashes if you have many tags. If you switch between tags to see the stats or while using the search, you can get "Page not work" error.

![Screenshot 2021-12-13 at 11 04 59](https://user-images.githubusercontent.com/39614855/145792254-496b7bd0-a4b3-4704-b857-2837f4533556.png)

My be it's a good idea to set the default mode to list with a default limit?

**Persist option Display Mode and Order by**
These two options resets if you leave the page for example to see stats of a tag. Is it possible to persist these options per session?

**Filter tags and export stats**
At the moment you have the chance to filter short-urls by tags with **or**. We would love to have the change to filter tags by **and**. Also it would be great to have a stats export after filtering the tags.

By the way, keep it up ;).

**Comment by acelaya** [comment]
The tags endpoint was always designed having in mind that you would usually never have so many, so it is not paginated (the pagination you can see in the web client is fully client-side, so it doesn't help with performance).

I will try to add support for pagination in the tags endpoint, in order to solve these performance issues https://github.com/shlinkio/shlink/issues/1273

> My be it's a good idea to set the default mode to list

This is already possible from the settings page. However, it won't help with the long loading times, because of the reasons mentioned above.

It might help with the crash though, as it will have to render way less components at a time.

> Persist option Display Mode and Order

As I was mentioning, the default mode can already be configured from the settings.

![Screenshot_20211214-074738](https://user-images.githubusercontent.com/2719332/145947207-6dd8e28e-c133-40d3-9157-af1b2506919f.png)

As per the default order, I'll tackle it separately https://github.com/shlinkio/shlink-web-client/issues/537

> At the moment you have the chance to filter short-urls by tags with or. We would love to have the change to filter tags by and.

This request makes sense, I have considered it myself many times. I have even found it counterintuitive.

I'll add a new argument to the short URLs endpoint that tells they must match **all** tags, not only **some** of them https://github.com/shlinkio/shlink/issues/1274

> Also it would be great to have a stats export after filtering the tags

~I'm not sure if I follow this one. What do you mean by "stats"? Do you mean being able to export the short URLs list after applying a filtering?~

Clarified in first reply. Tracked by https://github.com/shlinkio/shlink-web-client/issues/549

**Reply by tedo-develop** [reply]
> I'm not sure if I follow this one. What do you mean by "stats"? Do you mean being able to export the short URLs list after applying a filtering?

Yes, including the visits/clicks.

**Reply by acelaya** [reply]
The part about paginating the tags list has turned out to be muuuuch harder than I anticipated.

To begin with, after adding pagination I noticed the queries were still as slow as when loading the full list. In order to fix that I had to implement a new approach with nested queries.

However, this does not allow ordering the list by numbr of short URLs or visits. This is the only thing not yet supported, and it's going to be tricky.

**Reply by tedo-develop** [reply]
Hello Acelaya,

thanks for your reply. We hope you will solve this tricky problem.

**Reply by acelaya** [reply]
After trying all the database trickery I have been able to come-up with, I haven't been able to support paginated tags lists ordered by visits or short URLs in a performant way.

At the moment, the list can be paginated, and there's a clear performance gain if you order by tag name, even supporting filtering by text.

However, the aggregates are very tricky, and it will require some extra logic on top of it (most probably with some caching mechanism).

I will have to leave that out of the 3.0.0 milestone and tackle it sepparatly in #1346.

This means that, for the upcoming shlink-web-client 3.6.0 I have two options when consuming Shlink 3.0.0

* Don't change anything and continue loading the whole list, then do pagination + ordering + filtering on client-side
* Change the logic so that it does pagination + ordering + filtering in server-side when ordering by tag, and use existing approach only when ordering by tags or short-urls.

Ideally I would do the second, but it depends on how much complexity it adds, as it might not be worth it. Let's see.

Eventually, once #1346 has been implemented, shlink-web-client will move pagination + ordering + filtering to the server.

**Reply by tedo-develop** [reply]
Hello Acelaya,

we just want to say thanks for your lates updates. Our employees love the new features (export/tags filter).
It makes the work a lot easer!

**Reply by acelaya** [reply]
I love to hear that 😄
